state/cleanup.go

```diff
@@ -1433,6 +1436,62 @@ func (st *State) cleanupForceRemoveMachine(machineId string, cleanupArgs []bson.
 	return machine.Remove()
 }
 
+// cleanupEvacuateMachine is initiated by machine.Destroy() to gracefully remove units
+// from the machine before then kicking off machine destroy.
+func (st *State) cleanupEvacuateMachine(machineId string, cleanupArgs []bson.Raw) error {
+	if len(cleanupArgs) > 0 {
+		return errors.Errorf("expected no arguments, got %d", len(cleanupArgs))
+	}
+
+	machine, err := st.Machine(machineId)
+	if errors.IsNotFound(err) {
+		return nil
+	} else if err != nil {
+		return errors.Trace(err)
+	}
+	if machine.Life() != Alive {
+		return nil
+	}
+
+	units, err := machine.Units()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	if len(units) == 0 {
+		if err := machine.advanceLifecycle(Dying, false, false, 0); err != nil {
+			return errors.Trace(err)
+		}
+		return nil
+	}
+
+	buildTxn := func(attempt int) ([]txn.Op, error) {
+		if attempt > 0 {
+			units, err = machine.Units()
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+		}
+		var ops []txn.Op
+		for _, unit := range units {
+			destroyOp := unit.DestroyOperation()
+			op, err := destroyOp.Build(attempt)
+			if err != nil && !errors.Is(err, jujutxn.ErrNoOperations) {
+				return nil, errors.Trace(err)
+			}
+			ops = append(ops, op...)
+		}
+		return ops, nil
```

Comment:
If we get here, I can't see where we eventually call `m.advanceLifecycle(Dying, false, false, 0)`. Did I miss it?

Reply:
This just marks the unit as Dying, it first needs to transition to Dead, which will happen on a subsequent retry of the cleanup job